### PR TITLE
Use away-notify to show updates on users away state

### PR DESCRIPTION
**Pull request by MaxLeiter**
Probably need better wording for the settings option. Handles duplicate states (i.e. if someone goes away multiple times in a row it will only be shown once).

image

### CODE REVIEW

File: client/views/actions/away.tpl [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<span role="button" class="user {{colorClass from}}" data-name="{{from}}">{{mode}}{{from}}</span>
+{{#if text}}
+is now away
+<i class="away-message">({{{parse text}}})</i>
+{{else}}
+is back
```
Comment (MaxLeiter) [inline client/views/actions/away.tpl]: is now back?

File: client/index.html [hunk 2]
```diff
@@ -243,6 +243,12 @@ <h1 class="title">Settings</h1>
 									Show quits
 								</label>
 							</div>
+							<div class="col-sm-6">
+								<label class="opt">
+									<input type="checkbox" name="away">
+									Show aways/backs
```
Comment (MaxLeiter) [inline client/index.html]: Must be a better way to word this

File: src/models/user.js [hunk 3]
```diff
@@ -16,4 +17,5 @@ function User(attr, prefixLookup) {
 	// TODO: Remove this
 	this.name = this.nick;
 	this.mode = (this.modes && this.modes[0]) || "";
+	this.away = this.away;
```
Comment (MaxLeiter) [inline src/models/user.js]: Not necessary, just added for uniformity... should remove

File: client/index.html (hunk 2 above)
Comment (dgw) [inline client/index.html]: Ideas: "Show when users go away", "Show away notifications", "Notify on user presence changes", "Show user presence changes".

File: client/views/actions/away.tpl [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<span role="button" class="user {{colorClass from}}" data-name="{{from}}">{{mode}}{{from}}</span>
+{{#if text}}
+is now away
+<i class="away-message">({{{parse text}}})</i>
+{{else}}
+is now back
```
Comment (dgw) [inline client/views/actions/away.tpl]: I'm partial to wording it as "went away" and "came back", but whatever.

**Comment by astorije** [comment]
No matter what I try, multiple clients and users running this PR on 3 different IRC servers (including Freenode and the demo server), I cannot get to have anything showing up. Using `scripts/run-pr.sh` so the webpacking is done correctly.
Any idea? Any chance this PR could be currently broken? :/

File: client/index.html [hunk 5]
```diff
@@ -243,6 +243,12 @@ <h1 class="title">Settings</h1>
 									Show quits
 								</label>
 							</div>
+							<div class="col-sm-6">
+								<label class="opt">
+									<input type="checkbox" name="away">
+									Show when users go away
```
Comment (astorije) [inline client/index.html]: As opposed to @dgw's opinion, I think we should go with consistency, therefore something like `Show away/back changes` or the likes. It's not pretty, but it sticks out compared with the other options right now.

File: client/views/actions/away.tpl [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<span role="button" class="user {{colorClass from}}" data-name="{{from}}">{{mode}}{{from}}</span>
+{{#if text}}
+went away
+<i class="away-message">({{{parse text}}})</i>
+{{else}}
+came back
+{{/if}}
```
Comment (astorije) [inline client/views/actions/away.tpl]: `/nick` shows present tense (`x is now known as y`) so we could just go with `x is away` / `x is back` for consistency.

Also, I believe on The Lounge `/away` with no reason sets an away message of ` ` (one whitespace) so I think the message should now show up at all if trimmed `text` is empty.

File: src/plugins/irc-events/away.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+"use strict";
+
+let _ = require("lodash");
+let Msg = require("../../models/msg");
```
Comment (astorije) [inline src/plugins/irc-events/away.js]: `const` instead.

File: src/plugins/irc-events/away.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+module.exports = function(irc, network) {
+	let client = this;
+	irc.on("away", function(data) {
```
Comment (astorije) [inline src/plugins/irc-events/away.js]: Arrow function? :)

**Comment by MaxLeiter** [comment]
No idea why not working, will investigate. Regarding duplicate messages, people can't change their nick to the same nick, part twice, join twice, etc (without actions in between)

File: client/views/actions/away.tpl (hunk 6 above)
Comment (MaxLeiter) [inline client/views/actions/away.tpl]: Is there an easy way to tell what 'text' contains?

Comment (astorije) [inline client/views/actions/away.tpl]: I think you would have to define a helper for that which would do a stupid check like `text.trim().length === 0`. Maybe @xPaw would have a better idea?

Comment (xPaw) [inline client/views/actions/away.tpl]: Do we trim other messages like kicks? If not, don't do it here.

Comment (astorije) [inline client/views/actions/away.tpl]: I'm not saying we should trim `text`. I'm saying the check to know if there is a text should take this into account. This is because when using The Lounge, sending `/away` [will send a `" "` reason](https://github.com/thelounge/lounge/blob/0d8b58425eff9cc2a9f7468fcc678f8b2e1e7156/src/plugins/inputs/away.js#L7) so anyone else using The Lounge would see `nick is now away ( )`.
In such situation, it should not display the parens.

**Comment by MaxLeiter** [comment]
Addressed all changes except the empty parenthesis if no away message is supplied (@astorije)

**Comment by MaxLeiter** [comment]
@astorije even if these messages condensed, I still don't see a reason to show multiple away's from the same user - how about they're hidden if the message is the same, and shown if not?

**Comment by astorije** [comment]
Testing this on testnet.inspircd.org using 2 accounts, I cannot see status change when using `/away` with no argument or when using `/back`.
Using `/away foobar` does show `astorije is away (foobar)` correctly however.

**Comment by astorije** [comment]
@MaxLeiter, did you see my last message?

**Comment by dgw** [comment]
That's a mess of commits…

**Comment by YaManicKill** [comment]
So, what's the current state of this? Is it still not working for @astorije entirely?

**Comment by astorije** [comment]
Right, it wouldn't work at all on Freenode (but I believe this is normal as not supported there) and not working properly on that other network.
@MaxLeiter, do you think you can/want to work on this before v2.3.0 or should we remove it from the milestone? :(

**Comment by MaxLeiter** [comment]
Ill get it done

**Comment by astorije** [comment]
@MaxLeiter, ping on this? Would be really nice to put this behind us!

**Comment by astorije** [comment]
@MaxLeiter, any chance you could fix conflicts / wrap up this PR? :)

File: src/plugins/irc-events/away.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+"use strict";
+
+const _ = require("lodash");
+const Msg = require("../../models/msg");
+
+module.exports = function(irc, network) {
+	let client = this;
```
Comment (astorije) [inline src/plugins/irc-events/away.js]: To fix ESLint, use `const` here.

File: src/plugins/irc-events/away.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	irc.on("away", (data) => {
+		let away = false;
+		if (data.message) {
+			away = true;
+		}
+		network.channels.forEach((chan, index) => {
+			let user = _.find(chan.users, {name: data.nick});
```
Comment (astorije) [inline src/plugins/irc-events/away.js]: To fix ESLint, use `const` here.

File: src/plugins/irc-events/away.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+			if (typeof user === "undefined") {
+				return;
+			}
+			if (user.away === away) {
+				return;
+			}
+
+			let msg = new Msg({
```
Comment (astorije) [inline src/plugins/irc-events/away.js]: To fix ESLint, use `const` here.

**Comment by MaxLeiter** [comment]
@xPaw if you could take a look at this that'd be great -- I'm lost. Basically, the empty `/away` command should emit `/away <space>`, however -somewhere- that " " is being stripped so it's being received by the IRC server as `/away`, which is really `/back`.

**Comment by xPaw** [comment]
@MaxLeiter are you talking about you sending `/away` not receiving it? If so, why is that a problem for you in this particular PR?

**Comment by MaxLeiter** [comment]
I guess so, @xPaw, because for some reason `/away` no longer works with this PR.

File: src/plugins/irc-events/away.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+			const msg = new Msg({
+				type: Msg.Type.AWAY,
+				text: away,
+				time: data.time,
+				from: data.nick,
+				mode: chan.getMode(data.nick)
+			});
+
+			chan.pushMessage(client, msg);
+			if (index === network.channels.length - 1) { // Last iteration, let's update the users status
```
Comment (xPaw) [inline src/plugins/irc-events/away.js]: User object is a child of channel, you have to update it for every channel.

File: src/plugins/irc-events/away.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+const _ = require("lodash");
+const Msg = require("../../models/msg");
+
+module.exports = function(irc, network) {
+	const client = this;
+	irc.on("away", (data) => {
+		let away = null;
+
+		if (data.message) {
```
Comment (xPaw) [inline src/plugins/irc-events/away.js]: If `away.message` is an empty string here and you do want undefined:
```
const away = data.message || undefined;
```
otherwise just doing `const away = data.message;` is enough.

File: src/plugins/irc-events/away.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+		}
+
+		network.channels.forEach((chan, index) => {
+			const user = _.find(chan.users, {nick: data.nick});
+
+			if (typeof user === "undefined") {
+				return;
+			}
+
+			if (user.away === away) {
```
Comment (xPaw) [inline src/plugins/irc-events/away.js]: You can combine the ifs. `if (!user || user.away === away)`

File: src/plugins/inputs/away.js [hunk 15]
```diff
@@ -7,7 +7,6 @@ exports.input = function(network, chan, cmd, args) {
 
 	if (cmd === "away") {
 		reason = args.length > 0 ? args.join(" ") : " ";
-
```
Comment (xPaw) [inline src/plugins/inputs/away.js]: unnecessary change

**Comment by MaxLeiter** [comment]
Did some testing, PR is ready to go 👍

**Comment by astorije** [comment]
@xPaw, can you give the first review/approval to this please? Thanks!

**Comment by xPaw** [comment]
@astorije The code is fine, but I really don't know if we want to allow showing away changes like this.

**Comment by astorije** [comment]
Why is that? I think it's at least as valuable as nick changes, so I'm def okay with that personally.

File: client/index.html [hunk 16]
```diff
@@ -258,6 +258,12 @@ <h1 class="title">Settings</h1>
 									Show seconds in timestamp
 								</label>
 							</div>
+							<div class="col-sm-6">
+								<label class="opt">
+									<input type="checkbox" name="away">
+									Show away/back changes
```
Comment (YaManicKill) [inline client/index.html]: Should this be condensed when we do condensed messages? I reckon it should be just included like the rest of the join/part stuff.

**Comment by MaxLeiter** [comment]
Fine by me -- I'll rebase this once that's merged

**Comment by astorije** [comment]
@MaxLeiter, now that condensed stuff is out, do you think you'd be able to pick this up?
If not, @YaManicKill, do you think you could rebase/fix conflicts for this? As far as I can tell, all conflicts here are consequences of #759.

**Comment by YaManicKill** [comment]
The conflicts don't look that difficult, so if @MaxLeiter doesn't have time to do that, I'd be happy to fix the conflicts.

**Comment by astorije** [comment]
@YaManicKill, @MaxLeiter mentioned he wouldn't be able to take this up, so I assigned you to it, whenever you have a moment :)

File: client/index.html (hunk 16 above)
Comment (MaxLeiter) [reply client/index.html]: Condensed -- do we still want this option?

**Comment by MaxLeiter** [comment]
Rebased + updated to condense. Ready for review
image

Recommend testing on testnet.inspircd.org

File: client/index.html (hunk 16 above)
Comment (astorije) [reply client/index.html]: I don't think so. This is a status message like any others, I think. So something like "X users went away, Y users came back" and 💥, part of the condensed stuff.

Comment (astorije) [reply client/index.html]: Just noticed that's what you actually did so don't mind me 👍

**Comment by MaxLeiter** [comment]
Don't think we can control that, @astorije. I could make it show up but it wouldn't be too pretty -- I don't think it's necessary.

**Comment by astorije** [comment]
I think setting yourself as away/back should show, but since it would have nothing to do with away-notify, don't worry about it on this PR.

Glad we're actually going to merge this thing!! 😅

**Comment by astorije** [comment]
@YaManicKill or @xPaw, could one of you review this? :)

**Comment by xPaw** [comment]
@MaxLeiter Can you also mention away messages in help tooltip in the setting for status messages?

**Comment by astorije** [comment]
@MaxLeiter, could you squash?

**Comment by MaxLeiter** [comment]
Done

File: src/plugins/irc-events/away.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+			if (!user || user.away === away) {
+				return;
+			}
+
+			const msg = new Msg({
+				type: away ? Msg.Type.AWAY : Msg.Type.BACK,
+				text: away || "",
+				time: data.time,
+				from: data.nick,
+				mode: chan.getMode(data.nick)
```
Comment (xPaw) [inline src/plugins/irc-events/away.js]: You can do `user.mode` here, as you've already found the user.

**Comment by astorije** [comment]
Will merge once you address @xPaw's suggestion :)

**Comment by MaxLeiter** [comment]
Done
